### PR TITLE
ExPlat: Implement A/A tests

### PR DESCRIPTION
**Pull request by jessie-ross**
#### Changes proposed in this Pull Request

**This is a risky PR:** We are running a fairly large amount of new code in production in some fairly sensitive places.

This PR adds three A/A tests for the in-progress ExPlatJsClient:

* `explat_test_aa_calypso_boot`: `loadExperimentAssignment` in general calypso boot, run if logged in (so it won't run on the server), multiple calls with `useExperiment` and `dangerousGetExperimentAssignment` higher up.
* `explat_test_aa_calypso_signup`: `loadExperimentAssignment` in Calypso signup `index.web.ts`, with a `dangerousGetExperimentAssignment` higher up.
* `explat_test_aa_weekly_calypso_next_client_*`: Experiment component mirroring the existing evergreen A/A.

It aslo applies some minor fixes:
- Using `export type` and `import type`.
- Exporting `getTracksLoadPromise`.
- Changes `initializeAnonIdPromise` from a `const` to a `let` to fix an edge case.

#### Testing instructions

* Visit signed-in Calypso: ensure the `boot` experiment request is there.
* Visit signed-out Calypso: ensure the `signup` experiment request is there.
* Visit the Calypso Pages section: ensure the new evergreen aa experiment request is in there.
* Ensure there are no unusual console logged errors from the above.

Related to 585-gh-Automattic/experimentation-platform

### CODE REVIEW

File: packages/explat-client-react-helpers/src/index.tsx [hunk 1]
```diff
@@ -49,18 +48,11 @@ export default function createExPlatClientReactHelpers(
 			return () => {
 				isSubscribed = false;
 			};
-			// We don't expect experimentName to ever change and if it does we want to assignment to stay constant.
+			// experimentName can change for experiments with dates in their name, such as evergreen A/As
+			// but we don't want the assignment to change when it flips over, so we purposely don't include experimentName as a dep here.
 			// eslint-disable-next-line react-hooks/exhaustive-deps
 		}, [] );
 
-		if ( experimentName !== previousExperimentName ) {
```
Comment (yanirs) [inline packages/explat-client-react-helpers/src/index.tsx:56]: Maybe we should keep the check if both names don't start with `explat_test_aa_weekly_calypso_`? Dynamically created experiment names should be pretty rare.

Comment (jessie-ross) [reply packages/explat-client-react-helpers/src/index.tsx:56]: I am going to add a general exception for `explat_test_`

**Review by yanirs — APPROVED** [review]
LGTM, but I'm not familiar enough with Calypso to judge, so it'd probably be best to ask on the `#calypso` channel for a quick review before deploying. Should also verify that none of the build failures are due to this PR, and do some manual testing of affected pages/flows to be safe.

**Comment by jessie-ross** [comment]
> Here is how your PR affects size of JS and CSS bundles shipped to the user's browser:
>
> ```
> name                 parsed_size           gzip_size
> entry-main               +9110 B  (+0.7%)    +3118 B  (+0.9%)
> entry-login                +31 B  (+0.0%)       +9 B  (+0.0%)
> entry-gutenboarding        +31 B  (+0.0%)       +9 B  (+0.0%)
> ```

This is comparable to `/lib/abtest`:
> Screen Shot 2021-03-03 at 1 47 01 pm

(which will be removed in the future, leaving a net size increase of 1kb)

**Comment by jessie-ross** [comment]
I've also tested that this branch successfully builds the ETK

File: client/lib/explat/internals/anon-id.ts [hunk 2]
```diff
@@ -69,7 +71,7 @@ export const initializeAnonId = async (): Promise< string | null > => {
 				return;
 			}
 			attempt = attempt + 1;
-		}, anonIdPollingIntervalMilliseconds );
+		}, anonIdPollingIntervalMilliseconds ) as unknown ) as number;
```
Comment (jsnajdr) [inline client/lib/explat/internals/anon-id.ts]: Was ESLint or TypeScript complaining about anything here? When I revert this change, I don't see any errors.

In theory, `clearInterval( anonIdPollingInterval )` could be called before the `anonIdPollingInterval` value is assigned, but in practice this can never happen.

File: packages/explat-client/src/create-explat-client.ts [hunk 3]
```diff
@@ -43,6 +43,7 @@ export interface ExPlatClient {
 }
 
 export class MissingExperimentAssignmentError extends Error {
+	// eslint-disable-next-line @typescript-eslint/no-explicit-any
 	constructor( ...params: any[] ) {
```
Comment (jsnajdr) [inline packages/explat-client/src/create-explat-client.ts]: The `Error` constructor has just one well-defined parameter: `( message?: string )`. No need to assume arbitrary extra params here.

File: client/boot/common.js [hunk 4]
```diff
@@ -355,6 +356,11 @@ const setupMiddlewares = ( currentUser, reduxStore ) => {
 
 			next();
 		} );
+	} else {
+		// Temporary Experiment testing the new ExPlat client
+		loadExperimentAssignment( 'explat_test_aa_calypso_boot' );
+		// Duplicated on purpose
+		loadExperimentAssignment( 'explat_test_aa_calypso_boot' );
```
Comment (jsnajdr) [inline client/boot/common.js:363]: These calls can potentially return a rejected promise and trigger a console warning about unhandled promise rejections. Would be nice to wrap them with `catch( noop )` somehow.

Comment (jsnajdr) [reply client/boot/common.js:363]: Importing the ExPlat library during the boot also inflates the entrypoint code by 10kB which is quite significant. I assume that these tests are only temporary and will be removed after some time?

File: client/my-sites/customer-home/cards/tasks/site-setup-list/index.jsx [hunk 5]
```diff
@@ -36,6 +36,7 @@ import { getTask } from './get-task';
  * Style dependencies
  */
 import './style.scss';
+import { useExperiment } from 'calypso/lib/explat';
```
Comment (jsnajdr) [inline client/my-sites/customer-home/cards/tasks/site-setup-list/index.jsx]: Move this import to the "Internal dependencies" block.

File: client/my-sites/pages/main.jsx [hunk 6]
```diff
@@ -115,11 +116,22 @@ class PagesMain extends React.Component {
...
+				<Experiment
+					name={ `explat_test_aa_weekly_calypso_next_client_${ moment
+						.utc()
+						.format( 'GGGG' ) }_week_${ moment.utc().format( 'WW' ) }` }
+				>
+					{ {
+						treatment: null,
+						default: null,
+						loading: null,
+					} }
```
Comment (jsnajdr) [inline client/my-sites/pages/main.jsx:133]: Some feedback about the `<Experiment>` component API: it's very unusual that `children` is an object like this. A more classic way would be three props and no children:
```jsx
<Experiment
  treatment={ <div>treating</div> }
  default={ <div>defaulting</div> }
  loading={ <div>loading</div> }
/>
```

or a render prop that directly provides the info returned from `useExperiment`:
```jsx
<Experiment>
  { ( isLoading, assignment ) => {
    if ( isLoading ) {
      return <div>loading</div>;
    }
    if ( ! assignment?.variationName ) {
      return <div>defaulting</div>;
    }
    return <div>treating</div>;
  } }
</Experiment>
```
I would personally prefer this one. It makes the `<Experiment>` implementation an one-liner and the usage is not more complex at all.

The current `<Experiment>` has the disadvantage that the `treatment` branch doesn't have access to the `experimentAssignment` info at all. Does it really never need it?

File: packages/explat-client-react-helpers/src/index.tsx [hunk 7]
```diff
@@ -53,7 +53,10 @@ export default function createExPlatClientReactHelpers(
...
+		if (
+			experimentName !== previousExperimentName &&
+			! previousExperimentName.startsWith( 'explat_test' )
+		) {
 			const message = '[ExPlat] useExperiment: experimentName should never change between renders!';
```
Comment (jsnajdr) [inline packages/explat-client-react-helpers/src/index.tsx:60]: The `explat_test_aa_weekly_calypso` experiment is going to violate this constraint every week if you happen to be at the wrong Calypso screen at the wrong time 🙂 

The `useExperiment` hook could be rewritten to handle parameter name change gracefully. Let me know if you need any assistance with that.

**Review by jsnajdr — COMMENTED** [review]
This looks good :+1: Pointed out some places that should be improved a bit. Nothing serious.

File: client/lib/explat/internals/anon-id.ts (hunk 2 above)
Comment (jessie-ross) [reply client/lib/explat/internals/anon-id.ts]: I will play around with the typescript some more and see what is going on.

File: client/boot/common.js (hunk 4 above)
Comment (jessie-ross) [reply client/boot/common.js:363]: > These calls can potentially return a rejected promise and trigger a console warning about unhandled promise rejections. Would be nice to wrap them with catch( noop ) somehow.

`loadExperimentAssignment` is built to never throw in production:
https://github.com/Automattic/wp-calypso/blob/e57f7084bd853e364bfffdff7072d620eb5d0e95/packages/explat-client/src/create-explat-client.ts#L99-L166

Comment (jessie-ross) [reply client/boot/common.js:363]: > Importing the ExPlat library during the boot also inflates the entrypoint code by 10kB which is quite significant. I assume that these tests are only temporary and will be removed after some time?

So these tests are temporary, but it is likely for tests to be initialised here in the future. Importantly `/lib/abtest` adds about 6kb to `entry-main`, and will be removed in the near future (around a month pending developer buy-in), meaning only a net 3kb will be added.

File: client/my-sites/pages/main.jsx (hunk 6 above)
Comment (jessie-ross) [reply client/my-sites/pages/main.jsx:133]: > A more classic way would be three props and no children:

Thats a good point! I'll implement that.

> or a render prop that directly provides the info returned from useExperiment:

The idea here is that `<Experiment>` is safe and simple to use - it ensures that you have all three cases ready.

However I can see use for a `<WithExperiment>` component that does inject it in where hooks aren't available.

File: packages/explat-client-react-helpers/src/index.tsx (hunk 7 above)
Comment (jessie-ross) [reply packages/explat-client-react-helpers/src/index.tsx:60]: > The explat_test_aa_weekly_calypso experiment is going to violate this constraint every week if you happen to be at the wrong Calypso screen at the wrong time 🙂

The `! previousExperimentName.startsWith( 'explat_test' )` here should prevent that?

> The useExperiment hook could be rewritten to handle parameter name change gracefully. Let me know if you need any assistance with that.

The idea here is that once a component has rendered with an experiment, it should never change it's name, except for our internal automated A/A testing (tests we use to ensure the experiment system is not biased).

Comment (jsnajdr) [reply packages/explat-client-react-helpers/src/index.tsx:60]: I see, the `explat_test` prevents throwing the error :+1:

But still the `useEffect` hook that loads the assignment will not be run with the new (next week) name. It doesn't re-run on `experimentName` change. That's also desired in this case?

> The idea here is that once a component has rendered with an experiment, it should never change it's name

I know that in practice `useExperiment()` will always be called with the same `experimentName`. But it's a good practice in general for hooks to react correctly to parameter changes. It's similar to components reacting to a prop change, however unlikely. And the `useExperiment` hook could be easily adapted to make it work.

File: client/boot/common.js (hunk 4 above)
Comment (jessie-ross) [reply client/boot/common.js:363]: Hmm, I could also possibly remove the dependency on `@automattic/calypso-analytics`, bringing it almost in line with `/lib/abtest`:

Screen Shot 2021-03-03 at 7 55 51 pm

File: packages/explat-client-react-helpers/src/index.tsx (hunk 7 above)
Comment (jessie-ross) [reply packages/explat-client-react-helpers/src/index.tsx:60]: > But still the useEffect hook that loads the assignment will not be run with the new (next week) name. It doesn't re-run on experimentName change. That's also desired in this case?

Ah! Nice catch!

> I know that in practice useExperiment() will always be called with the same experimentName. But it's a good practice in general for hooks to react correctly to parameter changes. It's similar to components reacting to a prop change, however unlikely. And the useExperiment hook could be easily adapted to make it work.

Yeah I guess it should do so.

Comment (jessie-ross) [reply packages/explat-client-react-helpers/src/index.tsx:60]: > That's also desired in this case?

Actually, on second thought, I think this I did have this in mind - we should wait until the next time this component is constructed rather than switch experiment mid-render. I think that is a nice consistent rule to have for all experiments but other than that the two ways to go about it are the same.

File: client/boot/common.js (hunk 4 above)
Comment (jsnajdr) [reply client/boot/common.js:363]: > `loadExperimentAssignment` is built to never throw in production:

Thanks, I didn't notice that before :+1: So, if the `loadExperimentAssignment` REST request fails with a network error, a production instance will get some fallback assignment, while my development instance will break? The failure in the development mode seems a bit uncontrolled: for example, the `useEffect` callback in `useExperiment` will never set `isLoading` back to `false`.

Maybe even the development instance should get a fallback? It seems to me that the difference between production and development is now bigger that necessary.

Comment (jessie-ross) [reply client/boot/common.js:363]: 🤔 Maybe it should get a fallback. However I might push back on this for now and [create an issue](593-gh-Automattic/experimentation-platform) so we can proceed with the important A/A tests. The new client isn't released for general consumption and we can always change this behaviour later.

File: client/my-sites/pages/main.jsx (hunk 6 above)
Comment (jessie-ross) [reply client/my-sites/pages/main.jsx:133]: Issue for this created at 593-gh-Automattic/experimentation-platform, will be fixed after the A/A test.
